### PR TITLE
feat: calculate actual `receiptsRoot` when possible

### LINKED ISSUE #2667: Calculate actual receiptsRoot when possible
### Problem

Currently when we return block information, we don't include actual receiptsRoot, this is because Hedera is not using Merkle Patricia trie. However this can be easly calucalted during processing of the request. Returning real receiptsRoot can help tools, as well as help developers use merkle proofs of inclusion.

### Solution

Taking a look at how calculating receiptsRoot is done in the javascript execution client for ethereum can help us do it here as well. More information about the implementation can be found [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/src/buildBlock.ts#L155).

### PR DESCRIPTION
**Pull request by natanasow**
**Description**:

Currently, when we return block information, we don't include actual receiptsRoot, this is because Hedera is not using Merkle Patricia trie. However, this can be easily calculated during the processing of the request. Returning real `receiptsRoot` can help tools and developers.

**Solution**:

Taking a look at how calculating receiptsRoot is done in the javascript execution client for Ethereum can help us do it here as well. More information about the implementation can be found [here](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/src/buildBlock.ts#L155).

**Related issue(s)**:

Fixes #2667

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/relay/src/receiptsRootUtils.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      const topics: Uint8Array[] = [];
+      for (let j = 0; j < logs[i].topics.length; j++) {
+        topics.push(hexToBytes(logs[i].topics[j]));
+      }
+      // @ts-ignore
+      serializedLogs.push([hexToBytes(logs[i].address), topics, hexToBytes(logs[i].data)]);
+    }
+
+    return serializedLogs;
+  }
```
Comment (victor-yanev) [inline packages/relay/src/receiptsRootUtils.ts:41]:
```typescript
/**
 * Represents a serialized log entry as a tuple containing:
 * - The log address as a Uint8Array.
 * - The log topics as an array of Uint8Array.
 * - The log data as a Uint8Array.
 */
type SerializedLog = [Uint8Array, Uint8Array[], Uint8Array];

export class ReceiptsRootUtils {

  /**
   * Encodes an array of logs by converting them to SerializedLog tuples.
   *
   * @param logs - The array of logs to encode.
   * @returns An array of logs encoded as `SerializedLogEntry` tuples.
   */ 
  private static encodeLogs(logs: Log[]): SerializedLogEntry[] {
    const serializedLogs: SerializedLogEntry[] = [];
    
    for (let i = 0; i < logs.length; i++) {
      const topics: Uint8Array[] = [];
      for (let j = 0; j < logs[i].topics.length; j++) {
        topics.push(hexToBytes(logs[i].topics[j]));
      }
      serializedLogs.push([hexToBytes(logs[i].address), topics, hexToBytes(logs[i].data)]);
    }

    return serializedLogs;
  }
```

File: packages/relay/src/receiptsRootUtils.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+  private static encodeReceipt(receipt, txType) {
+    const encodedReceipt = RLP.encode([
+      receipt.root ?? (receipt.status === 0 ? Uint8Array.from([]) : hexToBytes('0x01')),
+      hexToBytes(receipt.cumulativeGasUsed),
+      hexToBytes(receipt.logsBloom),
+      this.encodeLogs(receipt.logs),
+    ]);
```
Comment (victor-yanev) [inline packages/relay/src/receiptsRootUtils.ts:49]: This is slightly easier to follow through:
```typescript
    let receiptRoot: Uint8Array;
    if (receipt.root) {
      receiptRoot = receipt.root;
    } else if (receipt.status === 0) {
      receiptRoot = Uint8Array.from([]);
    } else {
      receiptRoot = hexToBytes('0x01');
    }
    
    const encodedReceipt = RLP.encode([
      receiptRoot,
      hexToBytes(receipt.cumulativeGasUsed),
      hexToBytes(receipt.logsBloom),
      this.encodeLogs(receipt.logs),
    ]);
```

File: packages/relay/src/receiptsRootUtils.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        topics.push(hexToBytes(logs[i].topics[j]));
+      }
+      // @ts-ignore
+      serializedLogs.push([hexToBytes(logs[i].address), topics, hexToBytes(logs[i].data)]);
+    }
+
+    return serializedLogs;
+  }
+
+  private static encodeReceipt(receipt, txType) {
```
Comment (victor-yanev) [inline packages/relay/src/receiptsRootUtils.ts:43]: Please specify the argument types and the return type of this function

File: packages/relay/src/receiptsRootUtils.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    // legacy transactions
+    if (txType === 0) {
+      return encodedReceipt;
+    }
+
+    // EIP-2718 serialization
+    return concatBytes(intToBytes(txType), encodedReceipt);
+  }
+
+  public static async getRootHash(receipts) {
```
Comment (victor-yanev) [inline packages/relay/src/receiptsRootUtils.ts:60]: Same here

**Review by victor-yanev — CHANGES_REQUESTED** [review]

File: packages/relay/src/receiptsRootUtils.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+      return EthImpl.zeroHex32Byte;
+    }
+
+    const trie: Trie = new Trie();
+    receipts.map(async (receipt) => {
+      const path: Uint8Array =
+        receipt.transactionIndex === EthImpl.zeroHex
+          ? RLP.encode(Buffer.alloc(0))
+          : RLP.encode(bytesToInt(hexToBytes(receipt.transactionIndex ?? '')));
+      await trie.put(path, this.encodeReceipt(receipt, bytesToInt(hexToBytes(receipt.type ?? EthImpl.zeroHex))));
```
Comment (victor-yanev) [inline packages/relay/src/receiptsRootUtils.ts:82]: Currently it's quite hard to follow through what these lines aim to achieve:
```suggestion
      // add a comment explaining why we are doing this check (if variable is not self-sufficient)
      const variableExplainingTheCheck = receipt.transactionIndex === EthImpl.zeroHex;
      const path = variableExplainingTheCheck
          ? RLP.encode(Buffer.alloc(0))
          : RLP.encode(bytesToInt(hexToBytes(receipt.transactionIndex ?? '')));
      const txType = bytesToInt(hexToBytes(receipt.type ?? EthImpl.zeroHex));
      const encodedReceipt = this.encodeReceipt(receipt, txType);
      await trie.put(path, encodedReceipt);
```

File: packages/relay/src/receiptsRootUtils.ts [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    // legacy transactions
+    if (txType === 0) {
+      return encodedReceipt;
+    }
+
+    // EIP-2718 serialization
+    return concatBytes(intToBytes(txType), encodedReceipt);
+  }
+
+  public static async getRootHash(receipts: ITransactionReceipt[]): Promise<string> {
```
Comment (victor-yanev) [inline packages/relay/src/receiptsRootUtils.ts:71]: nit: public methods should be on top of private methods

**Review by victor-yanev — CHANGES_REQUESTED** [review]
LGTM! Few last suggestions:

**Review by victor-yanev — DISMISSED** [review]

**Review by victor-yanev — APPROVED** [review]
LGTM

**nit:** it would be nice if you could address the sonar issues about the for loop

**Review by konstantinabl — APPROVED** [review]

File: packages/relay/src/receiptsRootUtils.ts [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    // legacy transactions
+    if (txType === 0) {
+      return encodedReceipt;
+    }
+
+    // EIP-2718 serialization
+    return concatBytes(intToBytes(txType), encodedReceipt);
+  }
+
+  public static buildReceiptsFromTxHashesContractResultsAndLogs(
```
Comment (victor-yanev) [inline packages/relay/src/receiptsRootUtils.ts:93]:
```suggestion
  public static buildReceiptRootHashes(
```

**Review by victor-yanev — DISMISSED** [review]

**Review by isavov — DISMISSED** [review]

**Review by victor-yanev — APPROVED** [review]

**Review by ebadiere — APPROVED** [review]
LG.
